pkg/fleet-manager/fleet_plugin_metric.go

```diff
@@ -0,0 +1,339 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"kurator.dev/kurator/manifests"
+	fleetapi "kurator.dev/kurator/pkg/apis/fleet/v1alpha1"
+	"kurator.dev/kurator/pkg/fleet-manager/plugin"
+	"kurator.dev/kurator/pkg/infra/util"
+)
+
+func (f *FleetManager) reconcileObjstoreSecretOwnerReference(ctx context.Context, fleet *fleetapi.Fleet, fleetClusters map[string]*fleetCluster) error {
+	log := ctrl.LoggerFrom(ctx)
+	log = log.WithValues("fleet", types.NamespacedName{Name: fleet.Name, Namespace: fleet.Namespace})
+
+	for _, cluster := range fleet.Spec.Clusters {
+		fleetCluster, ok := fleetClusters[cluster.Name]
+		if !ok {
+			// should no happen
+			log.Error(nil, "failed to get cluster client")
+			continue
+		}
+
+		// reconcile objstore secret's owner reference
+		// a statefulset named prometheus-prometheus-prometheus is created by HelmRelease in each cluster
+		sts, err := fleetCluster.client.KubeClient().AppsV1().StatefulSets(MonitoringNamespace).Get(ctx, "prometheus-prometheus-prometheus", metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		secret, err := fleetCluster.client.KubeClient().CoreV1().Secrets(MonitoringNamespace).Get(ctx, fleet.Spec.Plugin.Metric.Thanos.ObjectStoreConfig.SecretName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		stsOwnerReference := metav1.OwnerReference{
+			APIVersion: "apps/v1",
+			Kind:       "StatefulSet", // TODO: use typemeta package
+			Name:       sts.Name,
+			UID:        sts.UID,
+		}
+		if !capiutil.HasOwnerRef(secret.OwnerReferences, stsOwnerReference) {
+			secret.OwnerReferences = append(secret.OwnerReferences, stsOwnerReference)
+			if _, err := fleetCluster.client.KubeClient().CoreV1().Secrets(MonitoringNamespace).Update(ctx, secret, metav1.UpdateOptions{}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// reconcileSidecarRemoteService reconciles a headless service named thanos-sidecar-remote, and ensure owner reference is set for all resources
+// TODO: find a better way to collect service endpoints of thanos-sidecar-remote service after all helm releases are reconciled
+func (f *FleetManager) reconcileSidecarRemoteService(ctx context.Context, fleet *fleetapi.Fleet, fleetClusters map[string]*fleetCluster) error {
+	log := ctrl.LoggerFrom(ctx)
+	log = log.WithValues("fleet", types.NamespacedName{Name: fleet.Name, Namespace: fleet.Namespace})
+
+	endpoints := sets.New[string]()
+	for _, cluster := range fleet.Spec.Clusters {
+		fleetCluster, ok := fleetClusters[cluster.Name]
+		if !ok {
+			// should no happen
+			log.Error(nil, "failed to get cluster client")
+			continue
+		}
+
+		svc, err := fleetCluster.client.KubeClient().CoreV1().Services(MonitoringNamespace).Get(ctx, PrometheusThanosServiceName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+			continue
+		}
+
+		for _, lb := range svc.Status.LoadBalancer.Ingress {
```

Comment:
There is a case when lb has not assigned an ip to the svc, we need to requeue

Reply:
that means, fleet will never became ready in the worest case.

Reply:
Yeah, that is fair point.
I am not sure how we can recover if the ingress ip not ready yet, is there max retry times.

Reply:
let me address this first, and fix with following PR